xwiki-platform-core/xwiki-platform-ckeditor/xwiki-platform-ckeditor-plugins/src/main/resources/xwiki-focusedplaceholder/plugin.js

```diff
@@ -0,0 +1,268 @@
+                    "i",
+                    "mark",
+                    "s",
+                    "small",
+                    "span",
+                    "strong",
+                    "time",
+                    "u",
+                    "var",
+                    "wbr",
+                    "del",
+                    "ins",
+                ],
+            };
+        },
+
+        init: function(editor) {
+
+            var focusedplaceholder = editor.plugins["xwiki-focusedplaceholder"];
+            focusedplaceholder.editor = editor;
+
+            // Create a TextWatcher instance to detect changes in the document
+            focusedplaceholder.placeholderTextWatcher = new CKEDITOR.plugins.textWatcher(editor,
+                function(range) {
+                    focusedplaceholder.updatePlaceholder(range);
+                });
+
+            // The placeholder should update at every keystroke
+            focusedplaceholder.placeholderTextWatcher.ignoredKeys = [];
+
+            editor.on('instanceReady', function() {
+                // Attach the placeholder to the editor
+                focusedplaceholder.placeholderTextWatcher.attach();
+
+                /* The placeholder should update when the content is changed via menus
+                 * snapshots are created when that occurs*/
+                editor.on("saveSnapshot", function() {
+                    focusedplaceholder.placeholderTextWatcher.check(false);
+                });
+            });
+
+
+            /* The placeholder should update when the user clicks somewhere in the document.
+             * Note: using contentDom event in case the DOM Tree gets recreated by CKEditor (e.g. mode change)*/
+            editor.on("contentDom", function() {
+                editor.document.$.addEventListener("click",
+                    function() {
+                        focusedplaceholder.placeholderTextWatcher.check(false);
+                    });
+            });
+        },
+
+        /**
+         * Removes any placeholder present in the document
+         *
+         * @param {Object} document - the DOM Document
+         */
+        removeAllPlaceholders: function(document) {
+            document.querySelectorAll("[" + ATTRIBUTE_NAME + "]").forEach(function(match) {
+                match.removeAttribute(ATTRIBUTE_NAME);
+            });
+        },
+
+        /**
+         * Checks if an element should be considered empty
+         *
+         * @param {Object} element - the DOM Element
+         * @return {bool}
+         */
+        isEmpty: function(element) {
+
+            // Consider the length of a text node as its number of children
+            if (element.nodeName === "#text") {
+                return !Boolean(element.length);
+            }
+            var i;
+            for (i = 0; i < element.childNodes.length; i++) {
+                var child = element.childNodes[i];
+                // Child is not empty if it doesn't have a nodeName
```

Comment:
Which nodes don't have a node name? Can you give an example?

Reply:
Ahah, I don't think I can give an example, I started with `tagName` so this was a check for non element nodes.
I removed the unnecessary check, thank you!